File: src/main/java/cspfj/constraint/semantic/AbsDiff.java
package cspfj.constraint.semantic;

import cspfj.constraint.AbstractAC3Constraint;
import cspfj.problem.Domain;
import cspfj.problem.Variable;

public final class AbsDiff extends AbstractAC3Constraint {

	public AbsDiff(final Variable result, final Variable v0, final Variable v1) {
		super(result, v0, v1);
	}

	@Override
	public boolean check() {
		return getValue(0) == Math.abs(getValue(1) - getValue(2));
	}

	@Override
	public boolean findSupport(final int variablePosition, final int index) {
		switch (variablePosition) {
		case 0:
			if (getVariable(1).getDomainSize() < getVariable(2).getDomainSize()) {
				return findValidTuple0(index, 1, 2);
			}
			return findValidTuple0(index, 2, 1);
		case 1:
			return findValidTuple(index, 1, 2);
		case 2:
			return findValidTuple(index, 2, 1);
		default:
			throw new IndexOutOfBoundsException();
		}
	}

	private boolean findValidTuple0(final int index, final int pos1,
			final int pos2) {
		final int val0 = getVariable(0).getValue(index);
		if (val0 < 0) {
			return false;
		}
		final Domain dom1 = getVariable(pos1).getDomain();
		final Domain dom2 = getVariable(pos2).getDomain();
		for (int i = dom1.firstIndex(); i >= 0; i = dom1.next(i)) {
			final int val1 = dom1.value(i);

			final int j1 = dom2.index(val1 - val0);
			if (j1 >= 0 && dom2.present(j1)) {
				tuple[0] = index;
				tuple[pos1] = i;
				tuple[pos2] = j1;
				return true;
			}

			final int j2 = dom2.index(val1 + val0);
			if (j2 >= 0 && dom2.present(j2)) {
				tuple[0] = index;
				tuple[pos1] = i;
				tuple[pos2] = j2;
				return true;
			}

		}
		return false;
	}

	private boolean findValidTuple(final int index, final int pos1,
			final int pos2) {
		final Domain result = getVariable(0).getDomain();
		final int val = getVariable(pos1).getValue(index);
		final Domain dom = getVariable(pos2).getDomain();
		for (int i = dom.firstIndex(); i >= 0; i = dom.next(i)) {
			final int resIndex = result.index(Math.abs(val - dom.value(i)));
			if (resIndex >= 0 && result.present(resIndex)) {
				tuple[0] = resIndex;
				tuple[pos1] = index;
				tuple[pos2] = i;
				return true;
			}
		}
		return false;
	}

	public String toString() {
		return getVariable(0) + " = |" + getVariable(1) + " - "
				+ getVariable(2) + "|";
	}

	@Override
	public float getEvaluation() {
		final int d0 = getVariable(0).getDomainSize();
		final int d1 = getVariable(1).getDomainSize();
		final int d2 = getVariable(2).getDomainSize();
		return d0 * d1 + d0 * d2 + d1 * d2;
	}
}

File: src/main/scala/cspfj/constraint/semantic/AbsDiff.scala
package cspfj.constraint.semantic;

import cspfj.constraint.Residues
import cspfj.problem.Domain
import cspfj.problem.Variable
import cspfj.constraint.Constraint
import cspfj.UNSATException
import cspfj.util.Interval

final class AbsDiff(val result: Variable, val v0: Variable, val v1: Variable)
  extends Constraint(Array(result, v0, v1)) with Residues {

  def checkValues(t: Array[Int]) = t(0) == math.abs(t(1) - t(2))

  def shave() = {
    val i0 = v0.dom.valueInterval
    val i1 = v1.dom.valueInterval

    val diff = i0 - i1

    var ch = result.dom.intersectVal(diff.abs)
    val r = result.dom.valueInterval

    if (diff.lb >= 0) {
      ch |= v0.dom.intersectVal(i1 + r)
      ch |= v1.dom.intersectVal(i0 - r)
    } else if (diff.ub <= 0) {
      ch |= v0.dom.intersectVal(i1 - r)
      ch |= v1.dom.intersectVal(i0 + r)
    } else {
      ch |= unionInter(v0.dom, i0, i1 + r, i0, i1 - r)
      ch |= unionInter(v1.dom, i1, i0 - r, i1, i0 + r)
    }

    ch
  }

  private def unionInter(dom: Domain, i0: Interval, j0: Interval, i1: Interval, j1: Interval) =
    (i0 intersect j0, i1 intersect j1) match {
      case (Some(k0), Some(k1)) => dom.intersectVal(k0 union k1) | (
        if (k0.ub < k1.lb) dom.filter(i => k0.ub >= dom.value(i) || dom.value(i) >= k1.lb)
        else if (k1.ub < k0.lb) dom.filter(i => k1.ub >= dom.value(i) || dom.value(i) >= k0.lb)
        else false)

      case (Some(k0), None) => dom.intersectVal(k0)
      case (None, Some(k1)) => dom.intersectVal(k1)
      case (None, None) => throw UNSATException.e
    }

  override def revise(mod: Seq[Int]) = {
    val ch = fixPoint(shave())

    if (isBound) {
      assert(boundConsistent, this + " is not BC")
      assert(!super.revise(mod), this + " is not BC")
      entailCheck(ch)
      ch
    } else super.revise(mod) || ch
  }

  override def findSupport(position: Int, index: Int) =
    position match {
      case 0 => findValidTuple0(index);
      case 1 => findValidTupleV1(index);
      case 2 => findValidTupleV2(index);
      case _ => throw new IndexOutOfBoundsException;
    }

  def findValidTuple0(index: Int) = {
    val val0 = scope(0).dom.value(index)
    if (val0 < 0) None
    else {
      val dom1 = scope(1).dom
      val dom2 = scope(2).dom
      dom1.indices map { i =>
        (i, dom2.index(dom1.value(i) - val0))
      } find {
        case (_, j) => j >= 0 && dom2.present(j)
      } orElse {
        dom1.indices map { i =>
          (i, dom2.index(dom1.value(i) + val0))
        } find {
          case (_, j) => j >= 0 && dom2.present(j)
        }
      } map {
        case (i, j) => Array(index, i, j)
      }
    }
  }

  def findValidTupleV1(index: Int): Option[Array[Int]] = {
    val result = this.result.dom
    val value = scope(1).dom.value(index);
    val dom = scope(2).dom;
    dom.indices map { i =>
      (i, result.index(math.abs(value - dom.value(i))))
    } find {
      case (_, resIndex) => resIndex >= 0 && result.present(resIndex)
    } map {
      case (i, resIndex) => Array(resIndex, index, i)
    }
  }

  def findValidTupleV2(index: Int): Option[Array[Int]] = {
    val result = this.result.dom
    val value = scope(2).dom.value(index);
    val dom = scope(1).dom;
    dom.indices map { i =>
      (i, result.index(math.abs(value - dom.value(i))))
    } find {
      case (_, resIndex) => resIndex >= 0 && result.present(resIndex)
    } map {
      case (i, resIndex) => Array(resIndex, i, index)
    }
  }

  override def toString = result + " = |" + v0 + " - " + v1 + "|";

  def getEvaluation = if (isBound) 5 else {
    val d0 = result.dom.size
    val d1 = v0.dom.size
    val d2 = v1.dom.size
    d0 * d1 + d0 * d2 + d1 * d2;
  }

  def simpleEvaluation = 2
}
